### PR TITLE
Gracefully handle errors for `download_file`

### LINKED ISSUE #45: `fuelup` should fail gracefully if binaries can't be found
Following on from #44, `fuelup` should fail gracefully if binaries for `forc*` and `fuel-core` can't be found.

### PR DESCRIPTION
**Pull request by binggh**
Closes #45 

Not done yet, think I want to implement a retry mechanism for failure scenarios (see: [ureq examples](https://docs.rs/ureq/latest/ureq/enum.Error.html#examples)) just leaving this up for visibility

### CODE REVIEW

File: src/download.rs [hunk 1]
```diff
@@ -91,19 +92,41 @@ fn unpack(tar_path: &Path, dst: &Path) -> Result<()> {
...
-    let mut data = Vec::new();
-    resp.into_reader().read_to_end(&mut data)?;
-
     let mut file = OpenOptions::new().write(true).create(true).open(&path)?;
-    if let Err(e) = file.write_all(&data) {
-        error!("Something went wrong writing to {}: {}", path.display(), e)
-    };
 
-    Ok(file)
+    for _ in 1..4 {
```
Comment (mitchmindtree) [inline src/download.rs:99]:
```suggestion
    for _ in 0..ATTEMPTS {
```

File: src/download.rs [hunk 2]
```diff
@@ -91,19 +92,41 @@ fn unpack(tar_path: &Path, dst: &Path) -> Result<()> {
     Ok(())
 }
 
-pub fn download_file(url: &str, path: &PathBuf) -> Result<File> {
+pub fn download_file(url: &str, path: &PathBuf) -> Result<(), anyhow::Error> {
```
Comment (mitchmindtree) [inline src/download.rs:95]:
```suggestion
pub fn download_file(url: &str, path: &Path) -> anyhow::Result<()> {
```

File: src/download.rs [hunk 3]
```diff
@@ -91,19 +92,41 @@ fn unpack(tar_path: &Path, dst: &Path) -> Result<()> {
...
+                        e
+                    )
+                };
+                return Ok(());
+            }
+            Err(ureq::Error::Status(404, r)) => {
+                // We've reached download_file stage, which means the tag must be correct.
+                error!("Failed to download from {}", &url);
+                let retry: Option<u64> = r.header("retry-after").and_then(|h| h.parse().ok());
+                let retry = retry.unwrap_or(3);
```
Comment (mitchmindtree) [inline src/download.rs:118]:
```suggestion
                let retry = retry.unwrap_or(RETRY_DELAY_SECS);
```

File: src/download.rs [hunk 4]
```diff
@@ -120,11 +143,10 @@ pub fn download_file_and_unpack(
...
     if download_file(&tarball_url, &tarball_path).is_err() {
-        error!(
-            "Failed to download from {} and write to path {}",
-            &tarball_url,
-            &tarball_path.display()
-        );
+        bail!(format!(
+            "Failed to download {} - the release might not be ready yet.",
+            &tarball_name
+        ));
```
Comment (mitchmindtree) [inline src/download.rs:149]: Remember that `bail!` also accepts literal format strings, e.g.

```rust
        bail!("Failed to download {} - the release might not be ready yet.", &tarball_name);
```

**Review by mitchmindtree — APPROVED** [review]

**Comment by binggh** [comment]
Merged changes from master. In master we already have `add_path_message` - previously I implemented `post_setup_message` ([9f69daf](https://github.com/FuelLabs/fuelup/pull/57/commits/9f69daf4f05aba3cf98cc471e89b59410a1fd819)) here but that's no longer needed.

Other than that the changes related to better error handling and retries should hold in this PR.

**Review by mitchmindtree — APPROVED** [review]
Looks good!

**Review by adlerjohn — APPROVED** [review]
utACK

**Pull request merged**